Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteGPUProxy.cpp

```diff
@@ -45,31 +46,30 @@ static constexpr size_t defaultStreamSize = 1 << 21;
 RemoteGPUProxy::RemoteGPUProxy(GPUProcessConnection& gpuProcessConnection, WebGPU::ConvertToBackingContext& convertToBackingContext, WebGPUIdentifier identifier, RenderingBackendIdentifier renderingBackend)
     : m_backing(identifier)
     , m_convertToBackingContext(convertToBackingContext)
-    , m_gpuProcessConnection(&gpuProcessConnection)
 {
     auto [clientConnection, serverConnectionHandle] = IPC::StreamClientConnection::create(defaultStreamSize);
     m_streamConnection = WTFMove(clientConnection);
-    m_gpuProcessConnection->addClient(*this);
-    m_gpuProcessConnection->connection().send(Messages::GPUConnectionToWebProcess::CreateRemoteGPU(identifier, renderingBackend, WTFMove(serverConnectionHandle)), 0, IPC::SendOption::DispatchMessageEvenWhenWaitingForSyncReply);
-    m_streamConnection->open(*this);
+    ensureOnMainRunLoop([identifier, renderingBackend, serverConnectionHandle = WTFMove(serverConnectionHandle)]() mutable {
+        WebProcess::singleton().ensureGPUProcessConnection().connection().send(Messages::GPUConnectionToWebProcess::CreateRemoteGPU(identifier, renderingBackend, WTFMove(serverConnectionHandle)), 0, IPC::SendOption::DispatchMessageEvenWhenWaitingForSyncReply);
+    });
+    m_streamConnection->open(*this, RunLoop::main());
     // TODO: We must wait until initialized, because at the moment we cannot receive IPC messages
     // during wait while in synchronous stream send. Should be fixed as part of https://bugs.webkit.org/show_bug.cgi?id=217211.
     waitUntilInitialized();
 }
 
 RemoteGPUProxy::~RemoteGPUProxy() = default;
 
-void RemoteGPUProxy::gpuProcessConnectionDidClose(GPUProcessConnection& connection)
+void RemoteGPUProxy::didClose(IPC::Connection&)
 {
-    ASSERT(m_gpuProcessConnection);
-    ASSERT(&connection == m_gpuProcessConnection);
+    ASSERT(m_streamConnection);
     abandonGPUProcess();
 }
 
 void RemoteGPUProxy::abandonGPUProcess()
 {
     m_streamConnection->invalidate();
-    m_gpuProcessConnection = nullptr;
+    m_streamConnection = nullptr;
```

Comment:
I think you cannot assign like this as the other webgpu objects might use the streamConnection() to obtain the connection and send to it.
Sending to invalid/closed connection is ok.
Currently we don't have a  sequence where we would guarantee no RemoteGPU**Proxy objects be accessed after losing the context. So it might be simpler to leave the stream connection ptr as is, for now.

Reply:
I'll move these changes to a separate issue, but do you think the solution here is to keep m_streamConnection, and implementing WebGL's `isContextLost()` using `m_streamConnection->isValid()` (which doesn't exist, but could).

Reply:
It perhaps could, but care should be taken to figure out the semantics. Naive/expected/most probable implementation of  `SCC::isValid()` would flip immediately when we know the other process crashes, but currently `RGCGLP::isContextLost()` will flip only due to the notification is being posted.

E.g. there could be a logical ordering bugs that then for example early-out omit sending the forceContextLost() upwards the inheritance chain